frontendUiRootImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/root/VcsRootComboBox.java

```diff
@@ -7,36 +7,61 @@
 import javax.swing.JComponent;
 import javax.swing.JPanel;
 
+import com.intellij.dvcs.repo.VcsRepositoryManager;
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.project.Project;
+import com.intellij.ui.GuiUtils;
 import com.intellij.ui.MutableCollectionComboBoxModel;
 import com.intellij.ui.SimpleListCellRenderer;
 import com.intellij.util.SmartList;
+import git4idea.GitUtil;
 import git4idea.repo.GitRepository;
 import io.vavr.collection.List;
 import io.vavr.control.Option;
 import org.checkerframework.checker.guieffect.qual.UIEffect;
-import org.checkerframework.common.value.qual.MinLen;
+import org.checkerframework.checker.nullness.qual.Nullable;
 
 import com.virtuslab.gitmachete.frontend.ui.api.root.IGitRepositorySelectionChangeObserver;
 import com.virtuslab.gitmachete.frontend.ui.api.root.IGitRepositorySelectionProvider;
+import com.virtuslab.logger.IPrefixedLambdaLogger;
+import com.virtuslab.logger.PrefixedLambdaLoggerFactory;
 
 public final class VcsRootComboBox extends JComboBox<GitRepository> implements IGitRepositorySelectionProvider {
 
+  private static final IPrefixedLambdaLogger LOG = PrefixedLambdaLoggerFactory.getLogger("frontendUiRoot");
+
   private final java.util.List<IGitRepositorySelectionChangeObserver> observers = new SmartList<>();
 
+  private final Project project;
+
   @UIEffect
-  public VcsRootComboBox(@MinLen(1) List<GitRepository> repositories) {
-    super(new MutableCollectionComboBoxModel<>(/* items */ repositories.asJavaMutable(), /* selection */ repositories.get(0)));
+  public VcsRootComboBox(Project project) {
+    super(new MutableCollectionComboBoxModel<>());
+    this.project = project;
+
+    updateRepositories();
     setRenderer(SimpleListCellRenderer.create( /* nullValue */ "", repo -> repo.getRoot().getName()));
+
+    project.getMessageBus().connect()
+        .subscribe(VcsRepositoryManager.VCS_REPOSITORY_MAPPING_UPDATED, () -> {
+          LOG.debug("VCS repository roots mappings changed");
+          GuiUtils.invokeLaterIfNeeded(() -> updateRepositories(), ModalityState.NON_MODAL);
+        });
   }
 
   @Override
   public MutableCollectionComboBoxModel<GitRepository> getModel() {
     return (MutableCollectionComboBoxModel<GitRepository>) super.getModel();
   }
 
-  @Override
   @UIEffect
-  public void updateRepositories(@MinLen(1) List<GitRepository> repositories) {
+  private void updateRepositories() {
+    // A bit of a shortcut: we're accessing filesystem even though we may be on UI thread here;
```

Comment:
`s/may be/are`

Reply:
If method is `@UIEffect`, then we can assume that it's invoked on UI thread *only*.
If method is `@SafeEffect` (i.e. non-UI-thread), then it may be invoked outside UI thread, but also on UI thread.